contracts/oracles/common/CappedOracle.sol

```diff
@@ -0,0 +1,120 @@
+
+    //// @notice Growth rate percentage in seconds. Ex: 1e18 is 100%
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable GROWTH_RATE_PER_SECOND;
+
+    /// @notice Snapshot update interval
+    /// @custom:oz-upgrades-unsafe-allow state-variable-immutable
+    uint256 public immutable SNAPSHOT_INTERVAL;
+
+    /// @notice Last stored snapshot price
+    uint256 public snapshotPrice;
+
+    /// @notice Last stored snapshot timestamp
+    uint256 public snapshotTimestamp;
+
+    /// @notice Emitted when the snapshot is updated
+    event SnapshotUpdated(uint256 price, uint256 timestamp);
+
+    /**
+     * @notice Constructor for the implementation contract.
+     * @param annualGrowthRate Annual growth rate in percentage
+     * @param snapshotInterval Snapshot update interval. snapshotInterval = 0 means caps are disabled
+     */
+    /// @custom:oz-upgrades-unsafe-allow constructor
+    constructor(uint256 annualGrowthRate, uint256 snapshotInterval) {
+        if (annualGrowthRate == 0 && snapshotInterval > 0) revert("Invalid growth rate");
+        SNAPSHOT_INTERVAL = snapshotInterval;
+        GROWTH_RATE_PER_SECOND = (annualGrowthRate) / (365 * 24 * 60 * 60);
+    }
+
+    /**
+     * @notice Returns if the price is capped
+     * @return isCapped Boolean indicating if the price is capped
+     */
+    function isCapped() external view virtual returns (bool) {
+        uint256 price = getUncappedPrice(token());
+
+        uint256 maxAllowedPrice = _getMaxAllowedPrice();
+
+        return (price > maxAllowedPrice) && (maxAllowedPrice != 0);
+    }
+
+    /**
+     * @notice Updates the snapshot price and timestamp
+     */
+    function updateSnapshot() public {
+        address asset = token();
+        if (Transient.readCachedPrice(CACHE_SLOT, asset) != 0) {
+            return;
+        }
+        if (block.timestamp - snapshotTimestamp < SNAPSHOT_INTERVAL || SNAPSHOT_INTERVAL == 0) return;
+
+        snapshotPrice = getPrice(asset);
+        snapshotTimestamp = block.timestamp;
+        Transient.cachePrice(CACHE_SLOT, asset, snapshotPrice);
+        emit SnapshotUpdated(snapshotPrice, snapshotTimestamp);
+    }
+
+    /**
+     * @notice Fetches the price of the token
+     * @param asset Address of the token
+     * @return price The price of the token in scaled decimal places. It can be capped
+     * to a maximum value taking into account the growth rate
+     */
+    function getPrice(address asset) public view override returns (uint256) {
+        uint256 price = Transient.readCachedPrice(CACHE_SLOT, asset);
+        if (price != 0) {
+            return price;
+        }
+
+        uint256 uncappedPrice = getUncappedPrice(asset);
+
+        if (SNAPSHOT_INTERVAL == 0) return uncappedPrice;
+
+        uint256 maxAllowedPrice = _getMaxAllowedPrice();
+
+        price = ((uncappedPrice > maxAllowedPrice) && (maxAllowedPrice != 0)) ? maxAllowedPrice : uncappedPrice;
```

Comment:
Open question: should we revert the transaction or return the capped price? AFAIK, aave returns the capped price